### PR TITLE
[Merged by Bors] - Initial Implementation of Sequences and Maps

### PR DESCRIPTION
**Pull request by JM4ier**
I added a sequence and a map type that translates to vipers sequence and map types
map is not finished at all
@vakaras to look over it

### CODE REVIEW

File: prusti-tests/tests/cargo_verify/veribetrfs/prusti-contracts [hunk 1]
```diff
@@ -0,0 +1 @@
+../../../../prusti-contracts
```
Comment (Aurel300) [inline prusti-tests/tests/cargo_verify/veribetrfs/prusti-contracts]: All of the `veribetrfs` symlinks are not meant for this PR, right?

File: prusti-viper/src/encoder/mir/types/encoder.rs [hunk 2]
```diff
@@ -109,10 +109,32 @@ impl<'p, 'v, 'r: 'v, 'tcx: 'v> TypeEncoder<'p, 'v, 'tcx> {
...
+            ty::TyKind::Adt(adt_def, substs) if adt_def.is_struct() => {
+                let type_name: &str = &self.encoder.env().tcx().def_path_str(adt_def.did());
+
+                let enc_substs = self
+                    .encode_substs(substs)
+                    .into_iter()
+                    .map(Box::new)
+                    .collect::<Vec<_>>();
+
+                // TODO: more stable reference to the Sequence type
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir/types/encoder.rs:122]: In the long-term I think it would be best if we marked all the Prusti-specific methods and types with an attribute like `prusti_builtin` (similar to what Rust does), then collect these into a list of "known" `DefId`s (similar to [`CommonTypes`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.CommonTypes.html)).

File: prusti-contracts/src/lib.rs [hunk 3]
```diff
@@ -47,6 +50,17 @@ mod private {
     /// A macro for defining a predicate using prusti expression syntax instead
     /// of just Rust expressions.
     pub use prusti_contracts_impl::predicate;
+
+
+    /// A sequence type
+    #[non_exhaustive]
+    pub struct Seq<T: Copy> {
```
Comment (Aurel300) [inline prusti-contracts/src/lib.rs]: This `Copy` requirement is a bit annoying. I think it roughly aligns with what we allow in pure functions but for things like ghost/model fields it is a bit limiting. E.g. if I wanted to store *copies* of intermediate states of a closure, I can only make this a `Copy` type by taking a shared reference, but this will then lead to borrow-checker issues further on. Since sequences and maps are spec-only anyway (I hope!), anything that is stored in a sequence can be snapshotted first, making it independent of the original data. Any thoughts @vakaras ?

File: prusti-tests/tests/cargo_verify/test_map/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+//
+extern crate prusti_contracts;
+use prusti_contracts::*;
```
Comment (vakaras) [inline prusti-tests/tests/cargo_verify/test_map/src/main.rs]: Could you please move these into the regular `https://github.com/viperproject/prusti-dev/tree/master/prusti-tests/tests/verify_overflow/fail/ghost`?

Comment (vakaras) [reply prusti-tests/tests/cargo_verify/test_map/src/main.rs]: After you removed the unnecessary files, could you please also squash the commits? I will then merge your PR.

**Comment by vakaras** [comment]
Regarding the failing CI:

* Formatting: `./x.py fmt-all`
* Clippy: `./x.py clippy --fix --allow-dirty` should automatically fix most of the warnings. You can then confirm this with `./x.py clippy -- -D warnings`.
* All tests: you need to add at least one `assert!()` that is expected to fail and mark the problematic line with a comment `//~ ERROR: <expected error message>`

File: prusti-contracts/src/lib.rs [hunk 5]
```diff
@@ -84,8 +98,59 @@ mod private {
...
+            panic!()
+        }
+        pub fn delete(self, _key: K) -> Self {
+            panic!()
+        }
+    }
+
+    pub macro ghost {
+        ($($t:tt)*) => {
+            $($t)*
```
Comment (Aurel300) [inline prusti-contracts/src/lib.rs]: Does this only take the first token or does it somehow implicitly use `tt` as well?

File: prusti-contracts/src/lib.rs [hunk 6]
```diff
@@ -84,8 +98,59 @@ mod private {
...
+        #[pure]
+        pub fn single(_: T) -> Self {
+            panic!()
+        }
+        #[pure]
+        pub fn concat(_: Self, _: Self) -> Self {
+            panic!()
+        }
+        #[pure]
+        pub fn lookup(self, index: usize) -> T {
```
Comment (Aurel300) [inline prusti-contracts/src/lib.rs]: Can we implement `Index<usize>` for `Seq` at least?

File: prusti-contracts/src/lib.rs [hunk 7]
```diff
@@ -84,8 +98,59 @@ mod private {
...
+
+    /// A map type
+    #[non_exhaustive]
+    #[derive(PartialEq, Eq, Copy, Clone)]
+    pub struct Map<K, V> {
+        _key_phantom: core::marker::PhantomData<K>,
+        _val_phantom: core::marker::PhantomData<V>,
+    }
+
+    impl<K, V> Map<K, V> {
```
Comment (Aurel300) [inline prusti-contracts/src/lib.rs:165]: `Index<K>` would be useful here as well.

File: prusti-tests/tests/cargo_verify/test_map/setup-links [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#! /bin/bash
```
Comment (Aurel300) [inline prusti-tests/tests/cargo_verify/test_map/setup-links]: Can we have just one copy of this script in `cargo_verify`? This is only useful for local testing, right? The script can take as argument a path to a subdirectory and set up the symlinks in that directory.

File: prusti-tests/tests/cargo_verify/test_map/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+//
```
Comment (Aurel300) [inline prusti-tests/tests/cargo_verify/test_map/src/main.rs]:
```suggestion
```

File: prusti-tests/tests/cargo_verify/test_map/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+extern crate prusti_contracts;
+use prusti_contracts::*;
+
+fn main() {
+    let m0: Map<u32, u32> = Map::empty();
+    let m1 = m0.insert(0, 0);
+    let m2 = m1.delete(0);
```
Comment (Aurel300) [inline prusti-tests/tests/cargo_verify/test_map/src/main.rs]: Although we don't check for this yet, this should be in `ghost! { ... }` as well. In the long term I think we need proper support for ghost arguments and return values but as we discussed with @vakaras and the others today, that's a bit tricky (I had [some ideas](https://hackmd.io/r0kiQvOMTdueQAZxKd4boQ?view#Ghost-arguments-and-extern_spec) but without patching the compiler like `verus` does we can't do that much).

File: prusti-tests/tests/cargo_verify/test_sequence/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+fn main() {
+    ghost! {
+        let s0: Seq<u32> = Seq::empty();
+        let one = 1;
+        let s1: Seq<u32> = Seq::single(one);
+        let s2: Seq<u32> = Seq::single(2);
+        let s3: Seq<u32> = Seq::concat(Seq::empty(), Seq::empty());
+        let s4: Seq<u32> = Seq::concat(s0, s1);
```
Comment (Aurel300) [inline prusti-tests/tests/cargo_verify/test_sequence/src/main.rs]: Add some assertions as well?

File: prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs [hunk 12]
```diff
@@ -0,0 +1,18 @@
+#[macro_use]
```
Comment (Aurel300) [inline prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs]: Is this required? I thought we can already use e.g. `predicate!` without the attribute.

File: prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+use prusti_contracts::*;
+
+#[requires(s.lookup(0) == 0)]
+fn foo(s: Seq<u32>) {
+
+}
+
+fn main() {
```
Comment (Aurel300) [inline prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs]: Why is this a failing test?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 14]
```diff
@@ -2844,6 +2845,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             }
         }
 
+        // checks if the function calls are operations on Sequences, and replaces these calls by Viper container Operations
+        if full_func_proc_name.starts_with("prusti_contracts::Seq::<T>::") {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Can we intercept this earlier? I.e. where we process terminators rather than making this a special case of an "impure" call? Also can it be a method of its own (`encode_seq_operation`)?

File: vir/defs/high/operations_internal/identifier/ty.rs [hunk 15]
```diff
@@ -37,6 +39,22 @@ impl WithIdentifier for ty::Int {
     }
 }
 
+impl WithIdentifier for ty::Sequence {
+    fn get_identifier(&self) -> String {
+        format!("seq${}", self.element_type.get_identifier())
```
Comment (Aurel300) [inline vir/defs/high/operations_internal/identifier/ty.rs]: Capital `Seq` since this is a type name?

File: vir/defs/high/operations_internal/identifier/ty.rs [hunk 16]
```diff
@@ -37,6 +39,22 @@ impl WithIdentifier for ty::Int {
...
+    }
+}
+
+impl WithIdentifier for ty::Map {
+    fn get_identifier(&self) -> String {
+        format!(
+            "map${}${}",
```
Comment (Aurel300) [inline vir/defs/high/operations_internal/identifier/ty.rs]: Capital?

File: vir/defs/polymorphic/ast/common.rs [hunk 17]
```diff
@@ -171,6 +171,16 @@ pub enum Type {
     TypeVar(TypeVar),
 }
 
+impl Type {
+    pub fn map(key_type: Type, val_type: Type) -> Type {
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/common.rs]: Why only for `map` and not `seq`?

File: vir/defs/polymorphic/ast/common.rs [hunk 18]
```diff
@@ -171,6 +171,16 @@ pub enum Type {
...
+        Type::Domain(DomainType {
+            label: "map".into(),
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/common.rs]: Capital?

File: vir/defs/polymorphic/ast/common.rs [hunk 19]
```diff
@@ -399,6 +413,10 @@ impl Type {
                     }
                 }
             }
+            Type::Seq(SeqType { box typ }) => {
+                // TODO is it alright to use `seq` as encoded name here?
+                format!("seq${}", Self::encode_arguments(&vec![typ.clone()]))
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/common.rs]: Capital?

File: x.py [hunk 20]
```diff
@@ -593,6 +600,9 @@ def main(argv):
             break
         else:
             cargo(argv[i:])
+            if arg == 'build':
```
Comment (Aurel300) [inline x.py]: This only creates the link *after* invoking `cargo`? In any case, I don't think these symlinks should be necessary at all, as long as you invoke the tool from `prusti-dev` as `target/debug/...`

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 21]
```diff
@@ -2712,6 +2712,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             .map(|arg| self.mir_encoder.encode_operand_place(arg))
             .collect::<Result<Vec<Option<vir::Expr>>, _>>()
             .with_span(call_site_span)?;
+
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
```

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 22]
```diff
@@ -2844,6 +2845,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+            let (encoded_target, mut stmts, typ, _) = self.encode_place(&place, ArrayAccessKind::Shared, location)?;
+            let position = self.register_error(call_site_span, ErrorCtxt::Unexpected);
+            let typ =  self.encoder.encode_type(typ).unwrap();
+
+            let encoded_args = arguments
+                .into_iter()
+                .map(|arg| vir::Expr::local(self.encode_prusti_local(arg)))
+                .collect::<Vec<_>>();
+
+            let rhs = match &full_func_proc_name[28..] {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: This `28..` is very fragile and not so obvious from the code. Match on the full proc names?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 23]
```diff
@@ -2844,6 +2845,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+                "empty" | "single" => vir::Expr::Seq(vir::Seq{
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
                "empty" | "single" => vir::Expr::Seq(vir::Seq {
```

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 24]
```diff
@@ -2844,6 +2845,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+                "concat" => vir::Expr::ContainerOp(vir::ContainerOp {
+                    op_kind: vir::ContainerOpKind::SeqConcat,
+                    left: Box::new(encoded_args[0].clone()),
+                    right: Box::new(encoded_args[1].clone()),
+                    position,
+                }),
+                _ => unreachable!("no further sequence functions."),
+            };
+
+            let ass = vir::Stmt::Assign(vir::Assign{
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
            let ass = vir::Stmt::Assign(vir::Assign {
```
Why the variable?

**Review by Aurel300 — COMMENTED** [review]
Added various comments… What about sets and multisets?

File: prusti-contracts/src/lib.rs (hunk 5 above)
Comment (JM4ier) [reply prusti-contracts/src/lib.rs]: `$t` captures all tokens that are given as arguments to the ghost macro, i.e. all token trees (tt)
I will rename it to avoid confusion.

File: prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs (hunk 12 above)
Comment (JM4ier) [reply prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs]: I am currently confused about the declarative macros, no amount of attributes seems to make them work.
I always get ```error: cannot find macro `ghost` in this scope```

File: x.py (hunk 20 above)
Comment (JM4ier) [reply x.py]: I added these symlinks to make prusti assistant work with the local version.
I can remove it if it's needed.

File: vir/defs/polymorphic/ast/common.rs (hunk 17 above)
Comment (JM4ier) [reply vir/defs/polymorphic/ast/common.rs]: I added the `map` function there to construct the domain type of the map, as I didn't see it suitable in other parts of the code.
I have since been informed that in newer versions of viper there is a builtin map type, which removes the need for the domain and this function.

File: prusti-tests/tests/cargo_verify/test_map/src/main.rs (hunk 4 above)
Comment (JM4ier) [reply prusti-tests/tests/cargo_verify/test_map/src/main.rs]: I cleaned it up a bit but the ghost macro still doesn't work which makes the tests not compile.

**Comment by vakaras** [comment]
@JM4ier Could you please fix CI?

File: prusti-tests/tests/verify_overflow/fail/core_proof/mathematical_types/sequences.rs [hunk 25]
```diff
@@ -0,0 +1,22 @@
+// compile-flags: -Punsafe_core_proof=true -Ponly_memory_safety=true
+
```
Comment (vakaras) [inline prusti-tests/tests/verify_overflow/fail/core_proof/mathematical_types/sequences.rs]: This file should be removed, right?

File: prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs [hunk 26]
```diff
@@ -0,0 +1,11 @@
+// compile-flags: -Punsafe_core_proof=true
+
+use prusti_contracts::*;
```
Comment (vakaras) [inline prusti-tests/tests/verify_overflow/fail/ghost/sequences.rs:3]: You must have at least one `//~ ERROR` in the test file.

File: prusti-common/src/vir/low_to_viper/ast.rs [hunk 27]
```diff
@@ -6,7 +6,7 @@ use vir::low::ast::{
     position::Position,
     predicate::PredicateDecl,
     statement::{self, Statement},
-    ty::{BitVector, BitVectorSize, Float, Type},
+    ty::{self as vir_ty, BitVector, BitVectorSize, Float, Type},
```
Comment (vakaras) [inline prusti-common/src/vir/low_to_viper/ast.rs]:
```suggestion
    ty::{Map, BitVector, BitVectorSize, Float, Type},
```

File: prusti-contracts/src/lib.rs [hunk 28]
```diff
@@ -1,4 +1,7 @@
 #![no_std]
+#![feature(register_tool)]
```
Comment (vakaras) [inline prusti-contracts/src/lib.rs]: These, I think, are not needed anymore, right?

File: prusti-contracts/src/lib.rs (hunk 7 above)
Comment (vakaras) [reply prusti-contracts/src/lib.rs:165]: Let's do this in a follow-up PR to avoid merge conflicts with modifications made by other students.

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs [hunk 29]
```diff
@@ -1461,6 +1468,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> BuiltinMethodsInterface for Lowerer<'p, 'v, 'tcx> {
                 vir_mid::TypeDecl::Array(_) => {
                     unimplemented!()
                 }
+                vir_mid::TypeDecl::Map(_) => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]: I think this should be next to `Sequence`.

File: prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs [hunk 30]
```diff
@@ -2015,6 +2026,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> BuiltinMethodsInterface for Lowerer<'p, 'v, 'tcx> {
                                 }
                             }
                             vir_mid::TypeDecl::Array(_) => unimplemented!("ty: {}", ty),
+                            vir_mid::TypeDecl::Map(_) => unimplemented!("ty: {}", ty),
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/builtin_methods/interface.rs]: Same comment: move `Map` next to `Sequence`.

File: prusti-viper/src/encoder/middle/core_proof/compute_address/interface.rs [hunk 31]
```diff
@@ -104,7 +104,8 @@ impl<'p, 'v: 'p, 'tcx: 'v> ComputeAddressInterface for Lowerer<'p, 'v, 'tcx> {
                 | vir_mid::TypeDecl::Int(_)
                 | vir_mid::TypeDecl::Float(_)
                 | vir_mid::TypeDecl::Pointer(_)
-                | vir_mid::TypeDecl::TypeVar(_) => {
+                | vir_mid::TypeDecl::TypeVar(_)
+                | vir_mid::TypeDecl::Sequence(_) => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/compute_address/interface.rs]: Add `Map`?

File: prusti-viper/src/encoder/middle/core_proof/snapshots/domains/interface.rs [hunk 32]
```diff
@@ -52,7 +58,17 @@ impl<'p, 'v: 'p, 'tcx: 'v> SnapshotDomainsInterface for Lowerer<'p, 'v, 'tcx> {
         &mut self,
         ty: &vir_mid::Type,
     ) -> SpannedEncodingResult<vir_low::Type> {
-        let domain_name = self.encode_snapshot_domain_name(ty)?;
-        self.domain_type(domain_name)
+        match ty {
+            vir_mid::Type::Map(map) => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/snapshots/domains/interface.rs:66]: Shouldn't `Sequence` also be here?

File: prusti-viper/src/encoder/middle/core_proof/types/interface.rs [hunk 33]
```diff
@@ -190,6 +193,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> Private for Lowerer<'p, 'v, 'tcx> {
                     )?;
                 }
             }
+            vir_mid::TypeDecl::Map(_decl) => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/types/interface.rs]: Could you merge `Map` and `Sequence` cases and add `unreachable!()`? (It should be guaranteed by the `if matches!` below.)

File: prusti-viper/src/encoder/middle/core_proof/utils/type_decl_encoder.rs [hunk 34]
```diff
@@ -20,7 +20,8 @@ pub(in super::super) trait TypeDeclWalker {
             | vir_mid::Type::Int(_)
             | vir_mid::Type::Float(_)
             | vir_mid::Type::Reference(_)
-            | vir_mid::Type::Pointer(_) => self.before_primitive(ty, parameters, lowerer),
+            | vir_mid::Type::Pointer(_)
+            | vir_mid::Type::Sequence(_) => self.before_primitive(ty, parameters, lowerer),
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/utils/type_decl_encoder.rs]: Also add `Map` here (also below).

File: prusti-contracts/src/lib.rs (hunk 28 above)
Comment (JM4ier) [reply prusti-contracts/src/lib.rs]: that was needed for the `#[pure]` annotations.
Since those didn't seem to work across crates anyways or had any use I removed it.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 35]
```diff
@@ -1281,6 +1284,27 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                     unimplemented!();
                 }
             }
+            "prusti_contracts::Int::new" => {
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:1287]: `Map` and `Sequence` functions are not here because they are currently supported only in contracts?

File: prusti-viper/src/encoder/mir/pure/interpreter/mod.rs [hunk 36]
```diff
@@ -436,109 +436,122 @@ impl<'p, 'v: 'p, 'tcx: 'v> ExpressionBackwardInterpreter<'p, 'v, 'tcx> {
...
+                    target_block,
+                    &states,
+                    &encoded_lhs,
+                    &encoded_args,
+                    span,
+                    substs,
+                )? {
+                    state
+                } else {
+                    match full_func_proc_name {
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/pure/interpreter/mod.rs]: All of these are also built-in functions. Could you move this match into `try_encode_builtin_funs` and if it makes it cleaner have your function `try_encode_builtin_type_function`.

File: prusti-viper/src/encoder/mir/pure/pure_functions/encoder.rs [hunk 37]
```diff
@@ -147,6 +147,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> PureFunctionEncoder<'p, 'v, 'tcx> {
         );
 
         let function_name = self.encode_function_name();
+
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/pure/pure_functions/encoder.rs]:
```suggestion

```

File: prusti-viper/src/encoder/mir/types/encoder.rs (hunk 2 above)
Comment (vakaras) [reply prusti-viper/src/encoder/mir/types/encoder.rs:122]: I like this idea. We should try this at some point.

File: prusti-contracts/src/lib.rs [hunk 38]
```diff
@@ -1,4 +1,5 @@
 #![no_std]
+#![feature(decl_macro)]
```
Comment (vakaras) [inline prusti-contracts/src/lib.rs]: Why is this needed? I do not see any macro declaration.
```suggestion
#![feature(decl_macro)]
```

**Review by vakaras — CHANGES_REQUESTED** [review]
Looks good in general. I added some small comments.

File: prusti-contracts/src/lib.rs (hunk 28 above)
Comment (vakaras) [reply prusti-contracts/src/lib.rs]: I see. Good.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs (hunk 35 above)
Comment (JM4ier) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:1287]: I only really focused on the contracts

Comment (vakaras) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:1287]: Ok, so that is what I expected.

**Review by vakaras — APPROVED** [review]
Looks good. Thanks!

bors r+

**Comment by vakaras** [comment]
@JM4ier Bors crashes when merging this PR because you did not set your email address in git:
```bash
$ git log -4
commit 8a46dfd81f0e6a90bd991083959e399f1d68a2c7 (HEAD, jonas-maier/sequences-maps)
Author: Jonas Maier <>
Date:   Wed May 11 22:56:25 2022 +0200

    fix formatting issues
```
Could you please fix this?

**Comment by JM4ier** [comment]
> @JM4ier Bors crashes when merging this PR because you did not set your email address in git:
> ```bash
> $ git log -4
> commit 8a46dfd81f0e6a90bd991083959e399f1d68a2c7 (HEAD, jonas-maier/sequences-maps)
> Author: Jonas Maier <>
> Date:   Wed May 11 22:56:25 2022 +0200
> 
>     fix formatting issues
> ```
> Could you please fix this?

Should I replace all commits that don't have my e-mail address or just add one new one that contains it?

**Comment by vakaras** [comment]
> > @JM4ier Bors crashes when merging this PR because you did not set your email address in git:
> > ```shell
> > $ git log -4
> > commit 8a46dfd81f0e6a90bd991083959e399f1d68a2c7 (HEAD, jonas-maier/sequences-maps)
> > Author: Jonas Maier <>
> > Date:   Wed May 11 22:56:25 2022 +0200
> > 
> >     fix formatting issues
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you please fix this?
> 
> Should I replace all commits that don't have my e-mail address or just add one new one that contains it?

Replace all.
